bokehjs/src/lib/core/util/templating.ts

```diff
@@ -3,84 +3,144 @@ import * as Numbro from "numbro"
 import tz = require("timezone")
 
 import {escape} from "./string"
-import {isNumber, isString} from "./types"
+import {isNumber, isString, isArray, isTypedArray} from "./types"
 
 import {ColumnarDataSource} from "models/sources/columnar_data_source"
+import {ImageIndex} from "../../models/glyphs/image"
 import {CustomJSHover} from 'models/tools/inspectors/customjs_hover'
 
-function _format_number(num: string | number): string {
-  if (isNumber(num)) {
+export type FormatterType = "numeral" | "printf" | "datetime"
+export type FormatterSpec = CustomJSHover | FormatterType
+export type Formatters = {[key: string]: FormatterSpec} | null
+export type FormatterFunc = (value:any, format:string, special_vars: Vars) => string
+export type Index = number | ImageIndex
+export type Vars = {[key: string]: any}
+
+export const DEFAULT_FORMATTERS = {
+  "numeral"  : function(value:any, format:string, _special_vars: Vars) { return Numbro.format(value, format) },
+  "datetime" : function(value:any, format:string, _special_vars: Vars) { return tz(value, format)            },
+  "printf"   : function(value:any, format:string, _special_vars: Vars) { return sprintf(format, value)       },
+}
+
+export function basic_formatter(value:any, _format:string, _special_vars: Vars): string {
+  if (isNumber(value)) {
     const format = (() => {
       switch (false) {
-        case Math.floor(num) != num:
+        case Math.floor(value) != value:
           return "%d"
-        case !(Math.abs(num) > 0.1) || !(Math.abs(num) < 1000):
+        case !(Math.abs(value) > 0.1) || !(Math.abs(value) < 1000):
           return "%0.3f"
         default:
           return "%0.3e"
       }
     })()
 
-    return sprintf(format, num)
-  } else
-    return `${num}`  // get strings for categorical types
+    return sprintf(format, value)
+  }
+
+  else
+    return `${value}`  // get strings for categorical types
+}
+
+export function get_formatter(name: string, raw_spec: string, format: string, formatters: Formatters = null) : FormatterFunc {
+  // no format, use default built in formatter
+  if (format==null)
+    return basic_formatter
+
+  // format spec in the formatters dict, use that
+  if (formatters != null && (name in formatters || raw_spec in formatters)) {
+
+    // some day (Bokeh 2.0) we can get rid of the check for name, and just check the raw spec
+    // keep it now for compatibility but do not demonstrate it anywhere
+    let key: string
+    if (raw_spec in formatters)
+      key = raw_spec
+    else
+      key = name
```

Comment:
Could be just `const key = raw_spec in formatters ? raw_spec : name`. Less vertical space wasted and no mutability allowed.

Reply:
Addressed in 6f2740d85